libraries/Microsoft.Bot.Builder/CloudAdapterBase.cs

```diff
@@ -0,0 +1,320 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// An adapter that implements the Bot Framework Protocol and can be hosted in different cloud environmens both public and private.
+    /// </summary>
+    public class CloudAdapterBase : BotAdapter
+    {
+        internal const string InvokeResponseKey = "BotFrameworkAdapter.InvokeResponse";
+
+        private ICloudEnvironment _cloudEnvironment;
```

Comment:
It seems that these 3 properties are read only, so it would be great to have const correctness

```suggestion
        private readonly ICloudEnvironment _cloudEnvironment;
```